### PR TITLE
DSL2: single-genome indexing

### PR DESCRIPTION
**Pull request by jfy133**
Adds indexing for when a user supplies a single fasta file for us as a reference.

I will do a separate PR for the input TSV.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](<https://github.com/>nf-core/eager/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by jfy133** [comment]
@nf-core-bot fix linting

**Comment by jfy133** [comment]
@nf-core-bot fix linting

File: conf/test.config [hunk 1]
```diff
@@ -22,8 +22,8 @@ params {
     // Input data
     // TODO nf-core: Specify the paths to your test data on nf-core/test-datasets
     // TODO nf-core: Give any required params for the test so that command line flags are not needed
-    input  = 'https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/samplesheet/samplesheet_test_illumina_amplicon.csv'
+    input = 'https://github.com/nf-core/test-datasets/raw/eager/testdata/Mammoth/samplesheet.tsv'
```
Comment (FriederikeHanssen) [inline conf/test.config:25]: does it actually work without the `raw.`...?

File: nextflow_schema.json [hunk 2]
```diff
@@ -63,6 +57,43 @@
...
+                "fasta_fai": {
+                    "type": "string",
+                    "default": "None",
+                    "description": "Path to samtools FASTA index (typically ending in '.fai'). If not supplied will be made for you.",
+                    "help_text": "If you want to use a pre-existing `samtools faidx` index, use this to specify the required FASTA index file for the selected reference genome. This should be generated by samtools faidx and has a file suffix of `.fai`."
+                },
+                "fasta_dict": {
+                    "type": "string",
+                    "default": "None",
+                    "description": "Path to picard sequence dictionary file (typically ending in '.dict'). If not supplied will be made for you..",
```
Comment (FriederikeHanssen) [inline nextflow_schema.json:69]:
```suggestion
                    "description": "Path to picard sequence dictionary file (typically ending in '.dict'). If not supplied will be made for you.",
```

File: nextflow_schema.json [hunk 3]
```diff
@@ -63,6 +57,43 @@
...
+                    "help_text": "If you want to use a pre-existing `picard CreateSequenceDictionary` dictionary file, use this to specify the required `.dict` file for the selected reference genome."
+                },
+                "fasta_mapperindexdir": {
+                    "type": "string",
+                    "default": "None",
+                    "description": "Path to directory containing index files of the FASTA for a given mapper..",
```
Comment (FriederikeHanssen) [inline nextflow_schema.json:75]:
```suggestion
                    "description": "Path to directory containing index files of the FASTA for a given mapper.",
```

File: subworkflows/local/reference_indexing_single.nf [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+
+    // TODO: document that the 'base name' of all indicies must be the same, i.e. correspond to the FASTA
+
+    // Join all together into a single map. Include failOnMismatch as a check if
+    // a user supplies indicies with different 'base' names.
+    ch_reference_for_mapping = ch_ungz_ref.join(ch_fasta_fai, failOnMismatch: true).join(ch_fasta_dict, failOnMismatch: true).join(ch_fasta_mapperindexdir, failOnMismatch: true)
+
+    emit:
+    reference = ch_reference_for_mapping // [ meta, fasta, fai, dict, mapindex ]
```
Comment (FriederikeHanssen) [inline subworkflows/local/reference_indexing_single.nf:77]: versions are not emitted

File: conf/test.config (hunk 1 above)
Comment (jfy133) [reply conf/test.config:25]: Yup! Apparently!

File: subworkflows/local/reference_indexing_single.nf (hunk 4 above)
Comment (jfy133) [reply subworkflows/local/reference_indexing_single.nf:84]: done in : https://github.com/nf-core/eager/pull/926/commits/996d101eb575e964628dba5657a5d25102ea0d5a

**Review by FriederikeHanssen — APPROVED** [review]
🚀

**Pull request merged**